supabase/migrations/30_manual_billing.sql

```diff
@@ -0,0 +1,354 @@
+            report
+        from internal.billing_report_202308(tenant_row.tenant_name, billed_month) as report
+        on conflict do nothing;
+
+        -- INSERT statements set FOUND true if at least one row is affected, false if no row is affected.
+        if found then
+          tenant_count = tenant_count + 1;
+        end if;
+    end loop;
+    return tenant_count;
+end
+$$ language plpgsql volatile;
+-- End internal billing report gen
+
+create or replace view invoices_ext as
+with has_bypassrls as (
+  select exists(select 1 from pg_roles where rolname = current_role and rolbypassrls = true) as bypass
+),
+authorized_tenants as (
+  select tenants.tenant
+  from tenants
+  left join has_bypassrls on true
+  left join auth_roles('admin') on tenants.tenant ^@ auth_roles.role_prefix
+  where has_bypassrls.bypass or auth_roles.role_prefix is not null
+),
+historical_bills as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz) as date_start,
+    date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day' as date_end,
+    report->>'billed_prefix' as billed_prefix,
+    report->'line_items' as line_items,
+    report->'subtotal' as subtotal,
+    report as extra
+  from billing_historicals
+  -- inner join should give only rows that match the join condition
+  inner join authorized_tenants on billing_historicals.tenant ^@ authorized_tenants.tenant
+),
+manual_bills as (
+  select
+    date_start,
+    date_end,
+    manual_bills.tenant as billed_prefix,
+    jsonb_build_array(
+      jsonb_build_object(
+        'description', manual_bills.description,
+        'count', 1,
+        'rate', manual_bills.usd_cents,
+        'subtotal', manual_bills.usd_cents
+      )
+    ) as line_items,
+    to_jsonb(usd_cents) as subtotal,
+    'null'::jsonb as extra
+  from manual_bills
+  inner join authorized_tenants on manual_bills.tenant ^@ authorized_tenants.tenant
+),
+current_month as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz) as date_start,
+    date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day' as date_end,
+    report->>'billed_prefix' as billed_prefix,
+    report->'line_items' as line_items,
+    report->'subtotal' as subtotal,
+    report as extra
+  from authorized_tenants, internal.billing_report_202308(authorized_tenants.tenant, now()) as report
+),
+combined as (
+  select
+    h.date_start, h.date_end, h.billed_prefix, h.line_items, h.subtotal, h.extra, 'usage' as invoice_type
+  from historical_bills h
+  union all
+  select
+    m.date_start, m.date_end, m.billed_prefix, m.line_items, m.subtotal, m.extra, 'manual' as invoice_type
+  from manual_bills m
+  union all
+  select
+    c.date_start, c.date_end, c.billed_prefix, c.line_items, c.subtotal, c.extra, 'current_month' as invoice_type
+  from current_month c
+)
+select * from combined
```

Comment:
i find it's general good practice not to put ordering constraints on views. Let the user of the view do that.